### PR TITLE
Add GitLab support

### PR DESCRIPTION
**Pull request by pauldub**
Hello,

This pull request adds initial support for Gitlab as requested in #13 #189 and #190.

I've based the implementation on the Github monitorable, and implemented the following tiles:

- GITLAB-COUNT
- GITLAB-MERGEREQUESTS
- GITLAB-PIPELINES

I'm not sure about how the pipeline's estimated duration should be computed and how the documentation is supposed to be updated.  Also the tests should be updated but I may have missed some of them.

I hope you find this work useful, and welcome your review on those changes.

### CODE REVIEW

**Comment by jsdidierlaurent** [comment]
Thanks a lot for your work !

I will watch what you did later. For now we are focus on fixing the most blocking problems.

See: https://github.com/monitoror/monitoror/milestone/4

File: service/server_api.go [hunk 1]
```diff
@@ -251,3 +259,26 @@ func (s *Server) registerGithub(variant string) {
 	s.configHelper.RegisterDynamicTileWithConfigVariant(github.GithubPullRequestTileType,
 		variant, &_githubModels.PullRequestParams{}, usecase)
 }
+
+func (s *Server) registerGitlab(variant string) {
+	gitlabConfig := s.config.Monitorable.Gitlab[variant]
+	// Custom UpstreamCacheExpiration only for count because gitlab has no-cache for this query and the rate limit is 30req/Hour
```
Comment (greut) [inline service/server_api.go]: Is this true as well?

https://docs.gitlab.com/ee/user/admin_area/settings/user_and_ip_rate_limits.html

Comment (pauldub) [reply service/server_api.go]: Indeed this is leftover from the GitHub module, I'll see to get it fixed. Another issue is that I found the global issues search api to be quite slow and hit the timeout frequently.

Comment (pauldub) [reply service/server_api.go]: According to https://gitlab.com/help/user/gitlab_com/index.md#gitlabcom-specific-rate-limits the gitlab.com rate limits are at 600 request/minute so we may not need a specific expiration.

I still need to investigate long response times from the search API.

**Comment by pauldub** [comment]
I've done more testing of this feature at work, and found that you quickly hit gitlab's rate limit of 10 request / sec / ip.

It may be possible to get a better user experience by implementing client side rate limiting, fair queuing of requests and more aggressive caching.

**Comment by Alex-D** [comment]
I do not know @jsdidierlaurent plans about this PR, but IMO a simple GitHub to GitLab copy paste is not enough. In all other monitorables, we take the time to fix all rate limit issues, timeout, etc. correctly. Also, some things that were needed on GitHub may not be useful for GitLab and vice-versa.

I think this kind of code can be a third-party monitorable in the future (see #253), but it does not hit the quality level that we want in the Core.

I know that the lack of "How to do a monitorable" documentation does not help, we will improve that through the next weeks/months (see #236).

--------

As an additional note to contribute to the rate limit issue, there is the current GitLab configuration: https://gitlab.com/gitlab-cookbooks/gitlab-haproxy/blob/master/attributes/default.rb.

> `default['gitlab-haproxy']['frontend']['api']['rate_limit_http_rate_per_minute'] = '600'`

> `default['gitlab-haproxy']['frontend']['https']['rate_limit_sessions_per_second'] = '10'`

But this is only for Gitlab.com, not for on-premise instances. Knowing that we need to make this an option.

**Comment by pauldub** [comment]
This is fine, I'm still getting the hang of how things are done, which is also why I've based the implementation on the GitHub module.

I'm willing to work towards improving the quality of the module. This process has prompted my previous comment about rate limits.

I'm not in a position to decide about wether or not this should be a third party monitorable, either way is fine by me.

I understand that on-premise instances need to be supported, but I would like to start with gitlab.com which is my primary use case. 

Anyway as I've said I'm currently still working on the implementation and welcome your opinion about this :)

**Comment by jsdidierlaurent** [comment]
Hello !

We refactored lot of code for improve core structure. 
So i create new PR with gitlab implementation: #340

Thank for you work ^^

**Pull request closed without merging**